### ISSUE TITLE
[WebAssembly] llvm-strip can invalidate linking sections

### ISSUE THREAD

**Issue by sbc100** [llvm/llvm-project#102002]
The WebAssembly version of llvm-strip currently just blindly copies all sections.

However, when stripping sections we really need to reconstruct the linking section and the reelections sections since symbol information can change.  This is what the ELF version of llvm-strip does I believe.

**Comment by sbc100** [comment]
See https://github.com/emscripten-core/emscripten/issues/22291

**Comment by ghostway0** [comment]
hello, I've started working on this a bit and I'm a bit puzzled (about the wasm spec): do all symbols appear in the linking section's symtab? if so, why are exports added to Symbols as well? wouldn't that be a duplicate entry? (see https://github.com/llvm/llvm-project/blob/main/llvm/lib/Object/WasmObjectFile.cpp#L1431)

**Comment by dschuff** [comment]
@sbc100 I updated the description to make it more clear that this only applies to stripping object files (although that is of course implied by the fact that only object files have linking sections), and also that it primary applies to the "strip-debug" use case since strip-all would of course also remove the linking section. Doing the work as described here would also open up the possibility of supporting some of the symbol-manipulation features of objcopy.

@ghostway0 Yes, for the linking use case, all symbols appear in the linking section's symtab. If a linking section is present, its contents [override](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Object/WasmObjectFile.cpp#L688) any information from the export section. Getting symbol information from the export or name section isn't used for linking, but it's useful for running tools like `llvm-nm` or objdump on linked files that aren't object files and have no symbol table. (This has several uses, but binary size analysis is probably the most common one.

**Comment by sbc100** [comment]
I also took a look at starting this yesterday.  It does look like it will be quite a large amount of work to make it work in the general case. 

For example, look at our current Wasm objcopy compared to the ELF version:

```
 $ wc -l llvm/lib/ObjCopy/ELF/ELFObj*
   992 llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp
  3078 llvm/lib/ObjCopy/ELF/ELFObject.cpp
  1246 llvm/lib/ObjCopy/ELF/ELFObject.h
  5316 total
$ wc -l llvm/lib/ObjCopy/wasm/Wasm*
  162 llvm/lib/ObjCopy/wasm/WasmObjcopy.cpp
   34 llvm/lib/ObjCopy/wasm/WasmObject.cpp
   48 llvm/lib/ObjCopy/wasm/WasmObject.h
   39 llvm/lib/ObjCopy/wasm/WasmReader.cpp
   31 llvm/lib/ObjCopy/wasm/WasmReader.h
   82 llvm/lib/ObjCopy/wasm/WasmWriter.cpp
   49 llvm/lib/ObjCopy/wasm/WasmWriter.h
  445 total
```

**Comment by dschuff** [comment]
Yeah, I'm not really surprised by this. Our objcopy implementation is extremely simple minimal.
It might not be as bad as the line count difference would suggest if we don't try to implement all the features of ELF objcopy (because there are a lot!). But having to parse all the symbols and relocations into some representation and then write them out again would certainly be nontrivial. How much "new" work it would take might possibly depend on how much of the related code that already does this in the linker (for reading) and backend/object writer (for writing) could be reused.

**Comment by ghostway0** [comment]
understood. didn't parse the clear there, for some reason. 
I looked at the code and the particular case of reconstructing the linking section sounds okay (the WasmObjectFile already gives us the symbols). relocations are associated with a WasmSection, so generating a relocations section for it might not be so bad after all (see https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/BinaryFormat/Wasm.h#L433)

I am of course not sure, and I'll work on it more tomorrow, if that's okay with you all

note that, as @dschuff mentioned, elf objcopy does so much more than the wasm version does

**Comment by ghostway0** [comment]
sat on this a bit today/yesterday. this is a _very_ rough and (currently) non-working, but I think this does get us closer :)
https://github.com/llvm/llvm-project/compare/main...ghostway0:llvm-project:ghostway/wasm-objcopy?expand=1

**Comment by sbc100** [comment]
Wow, thanks for working on this @ghostway0.   @dschuff and I discussed a stop-gap solution but it looks like you have the real thing almost working.

I worry that testing this stuff is going to be fairly involved too.

**Comment by ghostway0** [comment]
worked a bit more today and... I guess you're right. do you have any tools to decompile these segments and show problems that are not (printf) debugging? currently I'm debugging an `invalid function type` :)

**Comment by sbc100** [comment]
I've create a smaller, stopgap PR which fixes this bug without implementing full symbol table regeneration: #102978.  

@ghostway0 this is not intended as a replacement for your work, just something that we can land now while you work on the larger change.

**Comment by dschuff** [comment]
One possible way to try to test this kind of thing at a low level would be to leverage obj2yaml/yaml2obj to create and dump object files on the lowest level. So you could imagine a test written as a yaml file, that runs yaml2obj to "assemble" it, then llvm-strip, and then obj2yaml to check it, and the assertions would be written against the output yaml. This has the advantage that using yaml bypasses the object writer and reader and lets you create files that would not otherwise be valid. But it does involve a fair bit of manual work. Using the assembler and objdump is also an option (also useful for bootstrapping, e.g. you could write an assembly file, assemble it to an object file, and then dump that with ob2yaml to create some test input that you might later check in as yaml).

Another good option for debugging would be to use [wabt](https://github.com/webAssembly/wabt) and/or [wasm-tools](https://github.com/bytecodealliance/wasm-tools) to inspect your binaries at the lowest levels, which can give you different insight into the binary encodings.

**Comment by ghostway0** [comment]
using obj2yaml is a great idea!
I think I got it to work (stupid bug on my part). will clean it up and report back

**Comment by sbc100** [comment]
This issue has now been partially address with https://github.com/llvm/llvm-project/pull/102978, but I'll leave it open in case @ghostway0 is able to finish the more complete solution.

**Comment by ghostway0** [comment]
hi! sorry for the delay
worked on this today and I'm quite confused; I'm losing some bytes in the symtab (for example in `__memory_baseenv__table_base` and `sleep`) and getting some null section symbols, resulting in a `linking section ended prematurely`. I don't see where I could get them from (Symbols doesn't include them).

this is probably an unintended 'feature' I overlooked. will continue to debug it when I have time, and when that works, write the no-hack version :)
